File: reference/internal/azdevops/workitems.go
package azdevops

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// WorkItemTypeState represents a state available for a work item type
type WorkItemTypeState struct {
	Name     string `json:"name"`
	Color    string `json:"color"`
	Category string `json:"category"`
}

// WorkItemTypeStatesResponse represents the response from the work item type states API
type WorkItemTypeStatesResponse struct {
	Count int                 `json:"count"`
	Value []WorkItemTypeState `json:"value"`
}

// WorkItem represents a work item in Azure DevOps
type WorkItem struct {
	ID          int            `json:"id"`
	Rev         int            `json:"rev"`
	Fields      WorkItemFields `json:"fields"`
	URL         string         `json:"url"`
	ProjectName        string         `json:"-"` // Set by MultiClient, not from API
	ProjectDisplayName string         `json:"-"` // Set by MultiClient, display name for UI
}

// WorkItemFields represents the fields of a work item
type WorkItemFields struct {
	Title         string    `json:"System.Title"`
	State         string    `json:"System.State"`
	WorkItemType  string    `json:"System.WorkItemType"`
	AssignedTo    *Identity `json:"System.AssignedTo"`
	Priority      int       `json:"Microsoft.VSTS.Common.Priority"`
	ChangedDate   time.Time `json:"System.ChangedDate"`
	IterationPath string    `json:"System.IterationPath"`
	Description   string    `json:"System.Description"`
	ReproSteps    string    `json:"Microsoft.VSTS.TCM.ReproSteps"`
	Tags          string    `json:"System.Tags"`

	StoryPoints     float64   `json:"Microsoft.VSTS.Scheduling.StoryPoints"`
	StateChangeDate time.Time `json:"Microsoft.VSTS.Common.StateChangeDate"`
	ActivatedDate   time.Time `json:"Microsoft.VSTS.Common.ActivatedDate"`
	ClosedDate      time.Time `json:"Microsoft.VSTS.Common.ClosedDate"`
	CreatedDate     time.Time `json:"System.CreatedDate"`
}

// WorkItemReference represents a reference to a work item from WIQL queries
type WorkItemReference struct {
	ID  int    `json:"id"`
	URL string `json:"url"`
}

// WIQLResponse represents the response from a WIQL query
type WIQLResponse struct {
	WorkItems []WorkItemReference `json:"workItems"`
}

// WorkItemsResponse represents the response from getting work items
type WorkItemsResponse struct {
	Count int        `json:"count"`
	Value []WorkItem `json:"value"`
}

// StateIcon returns an icon for the work item state
// Workflow: New → Active → Resolved/Ready for Test → Closed
func (wi *WorkItem) StateIcon() string {
	stateLower := strings.ToLower(wi.Fields.State)

	switch {
	case stateLower == "new":
		return "○"
	case stateLower == "active":
		return "◐"
	case stateLower == "resolved" || strings.Contains(stateLower, "ready"):
		return "●"
	case stateLower == "closed":
		return "✓"
	case stateLower == "removed":
		return "✗"
	default:
		return "○"
	}
}

// EffectiveDescription returns the appropriate description field based on work item type.
// Bugs use Microsoft.VSTS.TCM.ReproSteps; other types use System.Description.
func (wi *WorkItem) EffectiveDescription() string {
	if wi.Fields.WorkItemType == "Bug" && wi.Fields.ReproSteps != "" {
		return wi.Fields.ReproSteps
	}
	return wi.Fields.Description
}

// TagList returns the tags as a trimmed slice, split on semicolons.
// Returns nil if there are no tags.
func (wi *WorkItem) TagList() []string {
	if wi.Fields.Tags == "" {
		return nil
	}
	raw := strings.Split(wi.Fields.Tags, ";")
	tags := make([]string, 0, len(raw))
	for _, t := range raw {
		t = strings.TrimSpace(t)
		if t != "" {
			tags = append(tags, t)
		}
	}
	return tags
}

// TimeInCurrentState reports how long the work item has been in its current state.
// Returns 0 when StateChangeDate is unset.
func (wi *WorkItem) TimeInCurrentState(now time.Time) time.Duration {
	if wi.Fields.StateChangeDate.IsZero() {
		return 0
	}
	return now.Sub(wi.Fields.StateChangeDate)
}

// EffectivePoints returns the work item's story-point estimate.
func (wi *WorkItem) EffectivePoints() float64 {
	return wi.Fields.StoryPoints
}

// IsCompletedSince reports whether the item is Closed and was closed strictly
// after `start`. Items with a zero ClosedDate or a non-Closed state return false.
func (wi *WorkItem) IsCompletedSince(start time.Time) bool {
	return strings.EqualFold(wi.Fields.State, "Closed") &&
		!wi.Fields.ClosedDate.IsZero() &&
		wi.Fields.ClosedDate.After(start)
}

// AssignedToName returns the display name of the assigned user, or "-" if unassigned
func (wi *WorkItem) AssignedToName() string {
	if wi.Fields.AssignedTo == nil {
		return "-"
	}
	return wi.Fields.AssignedTo.DisplayName
}

// QueryWorkItemIDs executes a WIQL query and returns the work item IDs
// top: maximum number of results to return
func (c *Client) QueryWorkItemIDs(query string, top int) ([]int, error) {
	path := fmt.Sprintf("/wit/wiql?api-version=7.1&$top=%d", top)

	payload := fmt.Sprintf(`{"query": %s}`, escapeJSONString(query))
	body, err := c.post(path, strings.NewReader(payload))
	if err != nil {
		return nil, fmt.Errorf("failed to execute WIQL query: %w", err)
	}

	var response WIQLResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, fmt.Errorf("failed to parse Azure DevOps API response for work item query: %w. "+
			"This may indicate an API structure change. Please check for updates or report this issue", err)
	}

	ids := make([]int, len(response.WorkItems))
	for i, wi := range response.WorkItems {
		ids[i] = wi.ID
	}

	return ids, nil
}

// GetWorkItems retrieves work items by their IDs
// Azure DevOps supports up to 200 IDs per request
func (c *Client) GetWorkItems(ids []int) ([]WorkItem, error) {
	if len(ids) == 0 {
		return []WorkItem{}, nil
	}

	// Convert IDs to comma-separated string
	idStrs := make([]string, len(ids))
	for i, id := range ids {
		idStrs[i] = strconv.Itoa(id)
	}
	idsParam := strings.Join(idStrs, ",")

	// Specify fields to retrieve
	fields := strings.Join([]string{
		"System.Id",
		"System.Title",
		"System.State",
		"System.WorkItemType",
		"System.AssignedTo",
		"Microsoft.VSTS.Common.Priority",
		"System.ChangedDate",
		"System.IterationPath",
		"System.Description",
		"Microsoft.VSTS.TCM.ReproSteps",
		"System.Tags",
		"Microsoft.VSTS.Scheduling.StoryPoints",
		"Microsoft.VSTS.Common.StateChangeDate",
		"Microsoft.VSTS.Common.ActivatedDate",
		"Microsoft.VSTS.Common.ClosedDate",
		"System.CreatedDate",
	}, ",")

	path := fmt.Sprintf("/wit/workitems?ids=%s&fields=%s&api-version=7.1", idsParam, fields)

	body, err := c.get(path)
	if err != nil {
		return nil, fmt.Errorf("failed to get work items: %w", err)
	}

	var response WorkItemsResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, fmt.Errorf("failed to parse Azure DevOps API response for work items: %w. "+
			"This may indicate an API structure change. Please check for updates or report this issue", err)
	}

	return response.Value, nil
}

// ListWorkItems retrieves work items assigned to the current user
// top: maximum number of work items to return (max 50 enforced)
func (c *Client) ListWorkItems(top int) ([]WorkItem, error) {
	// Enforce cap at 50
	if top > 50 {
		top = 50
	}

	// WIQL query to get active work items assigned to current user
	// @project scopes to the project in the API URL context, preventing
	// duplicates when multiple project clients query simultaneously
	query := `SELECT [System.Id] FROM WorkItems
WHERE [System.TeamProject] = @project
  AND [System.State] <> 'Closed'
  AND [System.State] <> 'Removed'
ORDER BY [System.ChangedDate] DESC`

	ids, err := c.QueryWorkItemIDs(query, top)
	if err != nil {
		return nil, err
	}

	if len(ids) == 0 {
		return []WorkItem{}, nil
	}

	return c.GetWorkItems(ids)
}

// ListMyWorkItems retrieves work items assigned to the authenticated user
// using the @Me WIQL macro, which Azure DevOps resolves server-side from the PAT.
// top: maximum number of work items to return (max 50 enforced)
func (c *Client) ListMyWorkItems(top int) ([]WorkItem, error) {
	if top > 50 {
		top = 50
	}

	query := `SELECT [System.Id] FROM WorkItems
WHERE [System.TeamProject] = @project
  AND [System.AssignedTo] = @Me
  AND [System.State] <> 'Closed'
  AND [System.State] <> 'Removed'
ORDER BY [System.ChangedDate] DESC`

	ids, err := c.QueryWorkItemIDs(query, top)
	if err != nil {
		return nil, err
	}

	if len(ids) == 0 {
		return []WorkItem{}, nil
	}

	return c.GetWorkItems(ids)
}

// GetWorkItemTypeStates retrieves the available states for a work item type.
// States in the "Removed" category are excluded since they are not typical user transitions.
func (c *Client) GetWorkItemTypeStates(workItemType string) ([]WorkItemTypeState, error) {
	path := fmt.Sprintf("/wit/workitemtypes/%s/states?api-version=7.1", workItemType)

	body, err := c.get(path)
	if err != nil {
		return nil, fmt.Errorf("failed to get work item type states: %w", err)
	}

	var response WorkItemTypeStatesResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, fmt.Errorf("failed to parse work item type states response: %w", err)
	}

	// Filter out "Removed" category states
	states := make([]WorkItemTypeState, 0, len(response.Value))
	for _, s := range response.Value {
		if s.Category != "Removed" {
			states = append(states, s)
		}
	}

	return states, nil
}

// UpdateWorkItemState updates the state of a work item using JSON Patch.
func (c *Client) UpdateWorkItemState(id int, state string) error {
	path := fmt.Sprintf("/wit/workitems/%d?api-version=7.1", id)

	payload := fmt.Sprintf(`[{"op":"replace","path":"/fields/System.State","value":%s}]`, escapeJSONString(state))
	_, err := c.doRequestWithContentType("PATCH", path, strings.NewReader(payload), "application/json-patch+json")
	if err != nil {
		return fmt.Errorf("failed to update work item state: %w", err)
	}

	return nil
}

File: src/Azdo.Core/AzureDevOps/WorkItems.cs
using System.Text.Json.Serialization;

namespace Azdo.Core.AzureDevOps;

/// <summary>A state available for a work item type.</summary>
public sealed class WorkItemTypeState
{
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("color")] public string Color { get; set; } = "";
    [JsonPropertyName("category")] public string Category { get; set; } = "";
}

/// <summary>Response from the work item type states API.</summary>
public sealed class WorkItemTypeStatesResponse
{
    [JsonPropertyName("count")] public int Count { get; set; }
    [JsonPropertyName("value")] public List<WorkItemTypeState> Value { get; set; } = new();
}

/// <summary>A work item in Azure DevOps.</summary>
public sealed class WorkItem
{
    [JsonPropertyName("id")] public int Id { get; set; }
    [JsonPropertyName("rev")] public int Rev { get; set; }
    [JsonPropertyName("fields")] public WorkItemFields Fields { get; set; } = new();
    [JsonPropertyName("url")] public string Url { get; set; } = "";

    /// <summary>Set by MultiClient, not from API.</summary>
    [JsonIgnore] public string ProjectName { get; set; } = "";
    /// <summary>Set by MultiClient, display name for UI.</summary>
    [JsonIgnore] public string ProjectDisplayName { get; set; } = "";

    /// <summary>An icon for the work item state. Workflow: New → Active → Resolved/Ready for Test → Closed.</summary>
    public string StateIcon()
    {
        string s = Fields.State.ToLowerInvariant();
        if (s == "new") return "○";
        if (s == "active") return "◐";
        if (s == "resolved" || s.Contains("ready")) return "●";
        if (s == "closed") return "✓";
        if (s == "removed") return "✗";
        return "○";
    }

    /// <summary>
    /// The appropriate description field based on work item type. Bugs use
    /// Microsoft.VSTS.TCM.ReproSteps; other types use System.Description.
    /// </summary>
    public string EffectiveDescription()
    {
        if (Fields.WorkItemType == "Bug" && !string.IsNullOrEmpty(Fields.ReproSteps))
            return Fields.ReproSteps;
        return Fields.Description;
    }

    /// <summary>The tags as a trimmed list split on semicolons; empty when there are no tags.</summary>
    public List<string> TagList()
    {
        if (string.IsNullOrEmpty(Fields.Tags))
            return new List<string>();
        var tags = new List<string>();
        foreach (var raw in Fields.Tags.Split(';'))
        {
            var t = raw.Trim();
            if (t.Length > 0)
                tags.Add(t);
        }
        return tags;
    }

    /// <summary>How long the work item has been in its current state. Zero when StateChangeDate is unset.</summary>
    public TimeSpan TimeInCurrentState(DateTime now)
    {
        if (Fields.StateChangeDate == default)
            return TimeSpan.Zero;
        return now - Fields.StateChangeDate;
    }

    /// <summary>The work item's story-point estimate.</summary>
    public double EffectivePoints() => Fields.StoryPoints;

    /// <summary>
    /// Whether the item is Closed and was closed strictly after <paramref name="start"/>.
    /// Items with a zero ClosedDate or a non-Closed state return false.
    /// </summary>
    public bool IsCompletedSince(DateTime start) =>
        string.Equals(Fields.State, "Closed", StringComparison.OrdinalIgnoreCase) &&
        Fields.ClosedDate != default &&
        Fields.ClosedDate > start;

    /// <summary>The display name of the assigned user, or "-" if unassigned.</summary>
    public string AssignedToName() => Fields.AssignedTo is null ? "-" : Fields.AssignedTo.DisplayName;
}

/// <summary>The fields of a work item.</summary>
public sealed class WorkItemFields
{
    [JsonPropertyName("System.Title")] public string Title { get; set; } = "";
    [JsonPropertyName("System.State")] public string State { get; set; } = "";
    [JsonPropertyName("System.WorkItemType")] public string WorkItemType { get; set; } = "";
    [JsonPropertyName("System.AssignedTo")] public Identity? AssignedTo { get; set; }
    [JsonPropertyName("Microsoft.VSTS.Common.Priority")] public int Priority { get; set; }
    [JsonPropertyName("System.ChangedDate")] public DateTime ChangedDate { get; set; }
    [JsonPropertyName("System.IterationPath")] public string IterationPath { get; set; } = "";
    [JsonPropertyName("System.Description")] public string Description { get; set; } = "";
    [JsonPropertyName("Microsoft.VSTS.TCM.ReproSteps")] public string ReproSteps { get; set; } = "";
    [JsonPropertyName("System.Tags")] public string Tags { get; set; } = "";

    [JsonPropertyName("Microsoft.VSTS.Scheduling.StoryPoints")] public double StoryPoints { get; set; }
    [JsonPropertyName("Microsoft.VSTS.Common.StateChangeDate")] public DateTime StateChangeDate { get; set; }
    [JsonPropertyName("Microsoft.VSTS.Common.ActivatedDate")] public DateTime ActivatedDate { get; set; }
    [JsonPropertyName("Microsoft.VSTS.Common.ClosedDate")] public DateTime ClosedDate { get; set; }
    [JsonPropertyName("System.CreatedDate")] public DateTime CreatedDate { get; set; }
}

/// <summary>A reference to a work item from WIQL queries.</summary>
public sealed class WorkItemReference
{
    [JsonPropertyName("id")] public int Id { get; set; }
    [JsonPropertyName("url")] public string Url { get; set; } = "";
}

/// <summary>Response from a WIQL query.</summary>
public sealed class WiqlResponse
{
    [JsonPropertyName("workItems")] public List<WorkItemReference> WorkItems { get; set; } = new();
}

/// <summary>Response from getting work items.</summary>
public sealed class WorkItemsResponse
{
    [JsonPropertyName("count")] public int Count { get; set; }
    [JsonPropertyName("value")] public List<WorkItem> Value { get; set; } = new();
}

/// <summary>A single comment from a work item's Discussion section.</summary>
public sealed class WorkItemComment
{
    [JsonPropertyName("id")] public int Id { get; set; }
    [JsonPropertyName("text")] public string Text { get; set; } = "";
    [JsonPropertyName("createdBy")] public Identity CreatedBy { get; set; } = new();
    [JsonPropertyName("createdDate")] public DateTime CreatedDate { get; set; }
}

/// <summary>CommentList wrapper returned by the work item comments GET endpoint.</summary>
public sealed class WorkItemCommentsResponse
{
    [JsonPropertyName("totalCount")] public int TotalCount { get; set; }
    [JsonPropertyName("count")] public int Count { get; set; }
    [JsonPropertyName("comments")] public List<WorkItemComment> Comments { get; set; } = new();
}

/// <summary>One state change on a work item, extracted from the /updates revision history.</summary>
public sealed class WorkItemStateTransition
{
    public string State { get; set; } = "";
    public DateTime At { get; set; }
}

/// <summary>
/// The configured state names used by the metrics WIQL. Mirrors the metrics
/// StateConfig so this layer doesn't depend on the metrics package. All three are required.
/// </summary>
public sealed class MetricsStateNames
{
    public string Active { get; set; } = "";
    public string ReadyForTest { get; set; } = "";
    public string Closed { get; set; } = "";
}
